drivers/encx24j600/encx24j600.c

```diff
@@ -0,0 +1,424 @@
+ *
+ * @}
+ */
+
+#include "mutex.h"
+#include "encx24j600.h"
+#include "encx24j600_internal.h"
+#include "encx24j600_defines.h"
+#include "vtimer.h"
+
+#include "net/dev_eth.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+#define ENCX24J600_SPI_SPEED SPI_SPEED_1MHZ
+
+#define ENCX24J600_INIT_DELAY 100000U
+
+#define ENC_BUFFER_START 0x0000
+#define ENC_BUFFER_SIZE  0x6000
+#define ENC_BUFFER_END   0x5FFF
+#define RX_BUFFER_START (0x5340) /* Default value */
+#define RX_BUFFER_END   (ENC_BUFFER_END)
+#define TX_BUFFER_LEN   (0x2000)
+#define TX_BUFFER_END   (RX_BUFFER_START)
+#define TX_BUFFER_START (TX_BUFFER_END - TX_BUFFER_LEN)
+
+static void receive_packets(encx24j600_t *dev);
+static void cmd(encx24j600_t *dev, char cmd);
+static void reg_set(encx24j600_t *dev, uint8_t reg, uint16_t value);
+static uint16_t reg_get(encx24j600_t *dev, uint8_t reg);
+static void reg_clear_bits(encx24j600_t *dev, uint8_t reg, uint16_t mask);
+
+/* dev_eth interface */
+static void _isr(dev_eth_t *dev);
+static int _init(dev_eth_t *dev);
+static int _send(dev_eth_t *dev, char *pkt, int len);
+static void _get_mac_addr(dev_eth_t *dev, uint8_t* buf);
+static int _recv(dev_eth_t *encdev, char* buf, int len);
+
+static inline int _get_promiscous(dev_eth_t *dev) {
+    return 1;
+}
+
+static inline int _set_promiscous(dev_eth_t *dev, int value) {
+    return 1;
+}
+
+const static eth_driver_t eth_driver_encx24j600 = {
+    .init = _init,
+    .send = _send,
+    .recv = _recv,
+    .get_mac_addr = _get_mac_addr,
+    .get_promiscous = _get_promiscous,
+    .set_promiscous = _set_promiscous,
+    .isr = _isr,
+};
+
+static inline void lock(encx24j600_t *dev) {
+    mutex_lock(&dev->mutex);
+}
+
+static inline void unlock(encx24j600_t *dev) {
+    mutex_unlock(&dev->mutex);
+}
+
+void encx24j600_setup(encx24j600_t *dev, spi_t spi, gpio_t cs, gpio_t int_pin)
+{
+    dev->ethdev.driver = &eth_driver_encx24j600;
+    dev->spi = spi;
+    dev->cs = cs;
+    dev->int_pin = int_pin;
+    dev->rx_next_ptr = RX_BUFFER_START;
+
+    mutex_init(&dev->mutex);
+}
+
+static void encx24j600_isr(void *arg)
+{
+    encx24j600_t *dev = (encx24j600_t *) arg;
+
+    /* disable interrupt line */
+    gpio_irq_disable(dev->int_pin);
+
+    /* call dev_eth hook */
+    dev_eth_isr((dev_eth_t*) dev);
+}
+
+static void _isr(dev_eth_t *encdev)
+{
+    encx24j600_t *dev = (encx24j600_t *) encdev;
+
+    uint16_t eir;
+
+    lock(dev);
+    cmd(dev, CLREIE);
+
+    eir = reg_get(dev, EIR);
+
+    if (eir & PKTIF) {
+        receive_packets(dev);
+    }
+
+    if (eir & LINKIF) {
+        uint16_t estat = reg_get(dev, ESTAT);
+        dev_eth_linkstate_handler(encdev, estat & PHYLNK);
+    }
+
+    /* drop all flags */
+    reg_clear_bits(dev, EIR, LINKIF);
+
+    /* re-enable interrupt */
+    gpio_irq_enable(dev->int_pin);
+    cmd(dev, SETEIE);
+
+    unlock(dev);
+}
+
+static inline void enc_spi_transfer(encx24j600_t *dev, char *out, char *in, int len)
+{
+    gpio_clear(dev->cs);
```

Comment:
you should lock the SPI bus before you use it -> `spi_lock()`. Other what happens if you share the bus between more than one device?